### ISSUE TITLE
Completely offload all initialization to 2nd Process

### ISSUE THREAD

**Issue by Xaymar** [Xaymar/OBS-AMD-Advanced-Media-Framework#378]
During OBS Studio startup the plugin currently tests if the encoder exists and doesn't crash in a second process, then tests again if it exists but this time in the main process. This results in a point-of-failure (POF) that should not exist, as it was already taken care of in a secondary process. 

Additionally all information gathering for GPUs can also be offloaded to this initial step, as it is very unlikely that a GPU suddenly appears while the operating system is still running. It might switch driver software, but the hardware ids should stay the same.

**Comment by Xaymar** [comment]
From Discord:

```
22:27] Jim: listen, we shouldn't initialize an encoder on startup
[22:27] Jim: I don't know what to do I admit
[22:27] Xaymar: Well, how else do you want to detect the amd encoder being available? What about hardware capabilities? Should I just suddenly list hundreds of options not actually supported?
[22:28] Xaymar: I had B-Frames listed for Polaris before I knew better. That resulted in 8 bug reports, several forum threads, and repeated users asking for help on their config :/
[22:28] Jim: with the NVENC plugin, the features are pretty limited.  I detect whether NVENC can work then blacklist known unsupported adapters by detecting their name
[22:28] Xaymar: Yeah that won't work with AMD.
[22:29] Jim: then you'll have to completely defer it in to the secondary test program
[22:29] Xaymar: To give you an example, R9 285 is R9 200 Series. Or R9 Series. Or R9 300 Series.
[22:29] Jim: because right now it's not doing that
[22:29] Jim: it only uses the test program to do the test to make sure it doesn't crash
[22:29] Xaymar: Yeah that's probably what I'll have to do. Completely offload all amd work to a 2nd program.
[22:29] Jim: not all work
[22:29] Xaymar: all detection and ui work
[22:30] Jim: just the program initialization
[22:30] Xaymar: ui initializes the encoder once to check supported options.
[22:30] Xaymar: Which I admit is awful
[22:30] Xaymar: But at the time I had to fix polaris
[22:31] Jim: wah?  listen anyway I don't know what's going on there but the initialization when the program starts up should be done in an entirely separate program due to this issue.  I'd do that with NVENC if I absolutely needed it but I just do the blacklist because it helps minimize startup startup time
[22:31] Jim: you do not have to defer it to a separate program for actually using it after the program starts up
[22:32] Jim: that'd probably be overkill
[22:32] Xaymar: To give you an overview of things, there are currently three initializations (for each encoder codec):
- OBS Start to detect hardware - should be offloaded to 2nd program.
- UI properties because there's no offline available settings, and also impossible to detect from name alone - can be offloaded to 2nd program.
- Actual encode.
[22:34] Jim: I'm not sure what you're talking about with UI properties, UI properties are only generated when they're actually used
[22:34] Jim: if I were in that position, I'd use the initialization part to get the information to generate the properties later
[22:35] Xaymar: That's probably the better way to do it.
[22:35] Jim: yea, I do this with game capture for example.  the plugins/win-capture/get-graphics-offsets thing for example
[22:35] Xaymar: Oh it's even worse than I thought, the encoder is tested every time the video api or adapter is changed.
[22:36] Xaymar: Literally can be offloaded to the detection part.
[22:36] Jim: the reason why is because initializing multiple directx versions at once in the program can cause a crash, so I have to aggregate that data in an offloaded process and then transfer it to OBS interprocess
[22:37] Jim: so this is something that we're not really a stranger to, it's just important to minimize unnecessary device initialization in OBS itself
[22:37] Jim: especially if it's just used to aggregate device information
[22:38] Jim: at least for these types of devices
[22:38] Xaymar: One question remains, how do we deal with process to process communication?
[22:38] Jim: even the win-dshow plugin has had this problem in the past with devices where certain devices could just crash OBS
[22:38] Jim: so it just sucks but we have to deal with it
[22:38] Xaymar: At least on my machines, the obs-ffmpeg-mux to obs communication randomly fails, resulting in a deadlocked obs that never quits. And I've also heard of this happening from FRANKIEonPCin1080p
[22:38] Jim: (thankfully it rarely happens with video devices)
[22:39] Jim: the failure of obs-ffmpeg-mux probably is not related, it depended up ffmpeg and people had random FFmpeg libraries on their system that were getting loaded instead and intefered
[22:40] Jim: where as obs-ffmpeg-mux explicitly must load the ones that came with obs
[22:40] Xaymar: Ah, no I have a completely different issue than that.
[22:40] Jim: that was likely the conflict for obs-ffmpeg-mux.  I'm fairly sure it's fixed with 23.2 already though
[22:40] Jim: because I moved it to the bin dir so it should prioritize the obs ffmpeg libraries
[22:40] Xaymar: obs-ffmpeg-mux deadlocks in the read function that obs provides, as the HANDLE it's trying to read from doesn't exist yet.
[22:41] Jim: you don't need to make it like obs-ffmpeg-mux, instead it should probably be more like get-graphics-offsets
[22:41] Jim: where it's a one-time transfer of information
[22:41] Jim: then terminates
[22:42] Jim: and it should really only transfer all the data one time (at startup).  if you need to look at the data again, my advice would probably be that you should just cache the enumeration data in obs after
[22:43] Xaymar: I think I only need data per GPU. Which I can probably cache without issues. I doubt gpus suddenly appear during runtime.
[22:43] Jim: at least that's my initial idea
[22:43] Jim: yea, exactly
```

**Comment by Xaymar** [comment]
Opting against this as there currently is no real use for such an advanced thing. It would be better to implement this on a plugin level to isolate entire plugins instead of just a single task.
